lib/rake/build.rake

```diff
@@ -96,6 +96,23 @@ namespace :build do
           RakeUtils.rake_stream_output 'seed:default', (rack_env?(:test) ? '--trace' : nil)
         end
 
+        # Restart Active Job workers before restarting dashboard server so that:
+        # 1. the order of the restarts will be consistent between production and
+        # other environments, and
+        # 2. the server code which is queueing new jobs does not need to be
+        # backward compatible (although the job code itself still does).
+        #
+        # When making breaking changes to a job's api contract, the best
+        # practice is to updated the job (in a backward compatible manner) in a
+        # first deploy, then update the code which calls it in a separate
+        # deploy, similarly to how we sequence deploys with database migrations
+        # or seeding changes.
+        #
+        # The sequencing described here is the best for mitigating any issues
+        # that may arise when that best practice is not followed.
+        ChatClient.log 'Restarting <b>dashboard</b> Active Job worker(s).'
+        RakeUtils.system 'bin/delayed_job', 'restart'
```

Comment:
Starting multiple jobs is easy to do if the number of jobs remains fixed:
```
Dave-MBP:~/src/cdo/dashboard (delayed-job-adapter)$ bin/delayed_job -n 2 restart
delayed_job.0: warning: no instances running. Starting...
delayed_job.0: process with pid 66780 started.
delayed_job.1: warning: no instances running. Starting...
delayed_job.1: process with pid 66782 started.
Dave-MBP:~/src/cdo/dashboard (delayed-job-adapter)$ ps -ef | grep delayed | grep -v grep
  501 66780     1   0  4:23PM ??         0:01.39 delayed_job.0
  501 66782     1   0  4:23PM ??         0:01.38 delayed_job.1
```
and it is easy to increase this number:
```
Dave-MBP:~/src/cdo/dashboard (delayed-job-adapter)$ bin/delayed_job -n 4 restart
delayed_job.0: trying to stop process with pid 66780 sending TERM and waiting 20s ...
delayed_job.0: process with pid 66780 successfully stopped.
delayed_job.0: process with pid 67006 started.
delayed_job.1: trying to stop process with pid 66782 sending TERM and waiting 20s ...
delayed_job.1: process with pid 66782 successfully stopped.
delayed_job.1: process with pid 67023 started.
delayed_job.2: warning: no instances running. Starting...
delayed_job.2: process with pid 67025 started.
delayed_job.3: warning: no instances running. Starting...
delayed_job.3: process with pid 67027 started.
Dave-MBP:~/src/cdo/dashboard (delayed-job-adapter)$ ps -ef | grep delayed | grep -v grep
  501 67006     1   0  4:25PM ??         0:01.33 delayed_job.0
  501 67023     1   0  4:25PM ??         0:00.47 delayed_job.1
  501 67025     1   0  4:25PM ??         0:00.36 delayed_job.2
  501 67027     1   0  4:25PM ??         0:00.36 delayed_job.3
```
but this naive approach leads to unacceptable behavior if the number of workers is later lowered:
```
Dave-MBP:~/src/cdo/dashboard (delayed-job-adapter)$ bin/delayed_job -n 3 restart
delayed_job.0: trying to stop process with pid 67006 sending TERM and waiting 20s ...
delayed_job.0: process with pid 67006 successfully stopped.
delayed_job.0: process with pid 67220 started.
delayed_job.1: trying to stop process with pid 67023 sending TERM and waiting 20s ...
delayed_job.1: process with pid 67023 successfully stopped.
delayed_job.1: process with pid 67238 started.
delayed_job.2: trying to stop process with pid 67025 sending TERM and waiting 20s ...
delayed_job.2: process with pid 67025 successfully stopped.
delayed_job.2: process with pid 67253 started.
Dave-MBP:~/src/cdo/dashboard (delayed-job-adapter)$ ps -ef | grep delayed | grep -v grep
  501 67027     1   0  4:25PM ??         0:12.15 delayed_job.3
  501 67220     1   0  4:26PM ??         0:02.32 delayed_job.0
  501 67238     1   0  4:26PM ??         0:01.32 delayed_job.1
  501 67253     1   0  4:26PM ??         0:00.47 delayed_job.2
```
not only is delayed_job.3 still running, but the PID is unchanged which means that it will be running old code indefinitely.

Reply:
full command line options can be found here: https://github.com/collectiveidea/delayed_job/blob/92fb435131d05b8a8715ac52acaacd9cde89619c/lib/delayed/command.rb#L28

Reply:
Is this script only ever executed on daemon servers?

I'm wondering if we should split this and "bin/delayed_job" out of this PR. The rest just sets us up to create jobs and add them to the database queue and I approve of those changes. My comments and questions are only in this build command change, and the "bin/delayed_job" script.

Reply:
> Is this script only ever executed on daemon servers?

yes, it's a long if clause starting here: https://github.com/code-dot-org/code-dot-org/blob/1fd334e357eef7b473110d9b1411bd66cc385ee0/lib/rake/build.rake#L58

> I'm wondering if we should split this and "bin/delayed_job" out of this PR. The rest just sets us up to create jobs and add them to the database queue and I approve of those changes. My comments and questions are only in this build command change, and the "bin/delayed_job" script.

Sure: https://github.com/code-dot-org/code-dot-org/pull/54021 Please keep in mind that bin/delayed_job is generated by `rails generate delayed_job:active_record` and its usage is referenced https://github.com/collectiveidea/delayed_job/blob/master/README.md, so we wouldn't want to change its behavior too drastically, though could see wanting to add comments or wrap it with certain protections in the prod environment or something like that.

Reply:
> I'm wondering if we should split this and "bin/delayed_job" out of this PR. The rest just sets us up to create jobs and add them to the database queue and I approve of those changes. My comments and questions are only in this build command change, and the "bin/delayed_job" script.
 
@cat5inthecradle any remaining concerns now that those things have been split out? if not, it would be great to get your approval on this and https://github.com/code-dot-org/code-dot-org/pull/53975 so that we can start chipping away at the stack of PRs here.

Reply:
This PR looks good to me, and I'm fine merging the other one into this one, then this one into staging. I'm dropping a note/question on the other PR though.

Reply:
Woohoo, thank you for the careful consideration here Darin!